modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1403 @@
+  return socketFd;
+}
+
+pragma "no doc"
+proc setBlocking(socketFd: fd_t, blocking: bool) throws {
+  var flags:c_int;
+  var err = sys_fcntl(socketFd, F_GETFL, flags);
+  if err != 0 {
+    throw SystemError.fromSyserr(err, "Failed to get socket flags");
+  }
+  if blocking {
+    flags &= ~O_NONBLOCK;
+  }
+  else {
+    flags |= O_NONBLOCK;
+  }
+  err = sys_fcntl_long(socketFd, F_SETFL, flags, flags);
+  if err != 0 {
+    throw SystemError.fromSyserr(err, "Failed to make socket non blocking");
+  }
+}
+
+pragma "no doc"
+proc bind(socketFd:fd_t, ref address: ipAddr, reuseAddr = true) throws {
+  var enable = (if reuseAddr then 1 else 0):c_int;
+  setSockOpt(socketFd, SOL_SOCKET, SO_REUSEADDR, enable);
+  var err = sys_bind(socketFd, address._addressStorage);
+  if err != 0 {
+    throw SystemError.fromSyserr(err, "Failed to bind Socket");
+  }
+}
+
+
+/*
+  Bind the socket to address. The socket must not already
+  be bound to any address prior to calling this procedure.
+
+  .. code-block:: Chapel
+
+    var socket =  new udpSocket();
+    var address = ipAddr.create("127.0.0.1", 8111);
+    bind(socket, address);
+
+  :arg socket: socket to set option on
+  :type socket: `tcpConn`
+  :arg address: address to bind to
+  :type address: `ipAddr`
+  :arg reuseAddr: whether to reuse address if already in use
+  :type address: `boolean`
+
+  :throws SystemError: If socket is closed or already bound.
+*/
+proc bind(ref socket: udpSocket, ref address: ipAddr, reuseAddr = true) throws {
+  var socketFd = socket.socketFd;
+  bind(socketFd, address, reuseAddr);
+}
+
+/*
+  See Above
+*/
+proc bind(ref socket: tcpListener, ref address: ipAddr, reuseAddr = true) throws {
+  var socketFd = socket.socketFd;
+  bind(socketFd, address, reuseAddr);
+}
+
+/*
+  See Above
+*/
+proc bind(ref socket: tcpConn, ref address: ipAddr, reuseAddr = true) throws {
+  var socketFd = socket.socketFd;
+  bind(socketFd, address, reuseAddr);
+}
+
+pragma "no doc"
+proc nagle(socketFd:fd_t, enable:bool) throws {
+  var c_enable = (if enable then 0 else 1):c_int;
+  setSockOpt(socketFd, IPPROTO_TCP, TCP_NODELAY, c_enable);
+}
+
+pragma "no doc"
+proc nagle(socketFd:fd_t):bool throws {
+  return if getSockOpt(socketFd, IPPROTO_TCP, TCP_NODELAY) == 0 then true else false;
+}
+
+/*
+  Enables or disables Nagle's algorithm on a given TCP Listener.
+
+  :arg enable: whether to enable or disable Nagle's algorithm
+  :type enable: `bool`
+
+  :throws SystemError: if not able to set `TCP_NODELAY` option properly.
+*/
+proc ref tcpListener.setNagle(enable:bool) throws {
+  var socketFd = this.socketFd;
+  nagle(socketFd, enable);
+}
+
+/*
+  Enables or disables Nagle's algorithm on a given TCP Connection.
+
+  :arg enable: whether to enable or disable Nagle's algorithm
+  :type enable: `bool`
+
+  :throws SystemError: if not able to set `TCP_NODELAY` flag properly.
+*/
+proc ref tcpConn.setNagle(enable:bool) throws {
+  var socketFd = this.socketFd;
+  nagle(socketFd, enable);
+}
```

Comment:
These methods are too far from the types that they are on. Why? Same goes for setDelayAck below.

Reply:
Earlier they were separate function the idea was to keep the documentation about Naggle optimization at one place only and keeping them at separate places will require us to add documentation at all three places(?) or we can reference the documentation for Naggle in other two 🤔

Reply:
Hmm, I don't know anything about this optimization, and it may make sense if there's such a context. I'll leave the call to you and @mppf

Reply:
I think it makes sense to have the methods on tcpConn appear in one place in the documentation. We can create a section on Nagle in the module-level docs comment to refer to from these.